java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/Application.java

```diff
@@ -205,4 +243,68 @@ private static void doForEachOverrideSetting(List<Field> configFields,
       }
     }
   }
+
+  /**
+   * Merges the given raw commandline settings with default settings for server and jdk images.
+   *
+   * @param rawServerSettings the raw commandline server mapping settings.
+   * @param rawJdkSettings the raw commandling jdk mapping settings.
+   * @return the merged settings.
+   */
+  @VisibleForTesting
+  public static JdkServerLookup mergeSettingsWithDefaults(String[] rawServerSettings,
```

Comment:
Is there any reason everything in this method cannot move to the `JdkServerLookup` class?
You could simply merge CLI and default settings before passing them into the `JdkServerLookup` constructor. You just need to make sure to maintain lookup precedence order.
This would eliminate the abstract class and template methods, which should generally be avoided. (see Item 18, in Effective Java 2nd Edition)

Reply:
Regarding not having `mergeSettingsWithDefaults` in `JdkServerLookup`:
The reason I didn't put it in there is because this function produces a version of JdkServerLookup that is kinda ad-hoc, and it's actually merging two JdkServerLookup objects to do so. It didn't make sense to me that JdkServerLookup objects would by default have a function to merge two JdkServerLookup objects in that very specific way. 

Regarding merging CLI and Default settings before passing to JdkServerLookup:
The reason I specifically avoided doing that is for the text validation stuff. For example, the user might input the same setting that looks very different because of spaces within the text or just one version of a setting might just be invalid. If i were to do this type of space-removing/trimming/validating then I would be replicating what's inside JdkServerLookup. 

Regarding not using abstract classes: I looked up item 18 but I think this book here is the wrong edition? Item 18 for me is about favoring static inner classes over nonstatic ones. However, I think there is a time and place for most java features, and I think in this situation the template method style is more elegant. I don't see what the difference wrt an interface, other than that I have factored out some common functionality into JdkServerLookup as an abstract class (vs it being an interface and having no common logic). I think the way I have it now very nicely ensures that both CLI and default settings are valid settings objects on their own, and that the ad-hoc merging of them in Application.java is also a valid settings object.

Reply:
continuing ... 

Given these requirements: 
1. CLI and default settings both need to be validated
2. CLI and default settings both need to be settings lookup objects on their own
3. A merge of the two settings needs to behave exactly the same way as the original CLI and default settings lookup objects

I think the template method style fits really well.